### PR TITLE
AuraDS size estimation tool

### PR DESCRIPTION
**Pull request by soerenreichardt**
~~Based on https://github.com/neo4j/graph-data-science-client/pull/629~~

### CODE REVIEW

**Comment by soerenreichardt** [comment]
[[🍓] Improve instance sizing API](https://trello.com/c/E6Au6njD/8926-%F0%9F%8D%93-improve-instance-sizing-api)

**Comment by soerenreichardt** [comment]
Is the v1 folder the correct place or does this need to go into v1beta4 or some other beta?

File: graphdatascience/session/aura_api.py [hunk 1]
```diff
@@ -247,6 +263,21 @@ def wait_for_instance_running(
...
+        self, node_count: int, relationship_count: int, algorithm_categories: List[AlgorithmCategory]
+    ) -> EstimationDetails:
+        data = {
+            "node_count": node_count,
+            "relationship_count": relationship_count,
+            "algorithm_categories": algorithm_categories,
+            "tier": "gds",  # fixed as professional or enterprise makes no difference in sizing
+        }
+
+        response = req.post(f"{self._base_uri}/v1/instances/estimations", headers=self._build_header(), json=data)
```
Comment (FlorentinD) [inline graphdatascience/session/aura_api.py:276]:
```suggestion
        response = req.post(f"{self._base_uri}/v1/instances/sizing", headers=self._build_header(), json=data)
```

File: graphdatascience/session/aura_api.py [hunk 2]
```diff
@@ -247,6 +263,21 @@ def wait_for_instance_running(
...
         return WaitResult.from_error(f"Instance is not running after waiting for {waited_time} seconds")
 
+    def estimate_size(
+        self, node_count: int, relationship_count: int, algorithm_categories: List[AlgorithmCategory]
+    ) -> EstimationDetails:
+        data = {
+            "node_count": node_count,
+            "relationship_count": relationship_count,
+            "algorithm_categories": algorithm_categories,
+            "tier": "gds",  # fixed as professional or enterprise makes no difference in sizing
```
Comment (FlorentinD) [inline graphdatascience/session/aura_api.py:273]:
```suggestion
            "instance_type": "gds",  # fixed as professional or enterprise makes no difference in sizing
```

File: graphdatascience/session/gds_sessions.py [hunk 3]
```diff
@@ -82,6 +84,23 @@ def __init__(self, api_credentials: AuraAPICredentials) -> None:
             client_secret=api_credentials.client_secret,
         )
 
+    def estimate(
+        self, node_count: int, relationship_count: int, algorithm_categories: Optional[List[AlgorithmCategory]] = None
+    ) -> SessionSizeByMemory:
+        if algorithm_categories is None:
+            algorithm_categories = []
```
Comment (soerenreichardt) [inline graphdatascience/session/gds_sessions.py:91]: We need to double check, but I think when algorithm categories is empty, the result is always to use 8gb

Comment (FlorentinD) [reply graphdatascience/session/gds_sessions.py:91]: good point! is that an EstimationCLI problem?
Shouldnt we at least estimate the projection based on the counts?

Comment (soerenreichardt) [reply graphdatascience/session/gds_sessions.py:91]: Not sure where this problem originates, either in console-api or the estimation cli.
Ideally the projection should be taken into account but as far as I recall that wasn't the case

File: graphdatascience/tests/unit/test_gds_sessions.py [hunk 4]
```diff
@@ -424,5 +435,25 @@ def test_invalid_session_name() -> None:
...
+def test_estimate_size_exceeds(aura_api: AuraApi) -> None:
+    aura_api = FakeAuraApi(size_estimation=EstimationDetails("16GB", "8GB", True))
+    sessions = GdsSessions(AuraAPICredentials("", "", "foo"))
+    sessions._aura_api = aura_api
+
+    with pytest.warns(
+        ResourceWarning,
+        match=re.escape("The estimated memory `16GB` exceeds the maximum size supported of your Aura tenant (`8GB`)"),
+    ):
+        sessions.estimate(1, 1, [AlgorithmCategory.CENTRALITY]) == SessionSizeByMemory.M
```
Comment (Mats-SX) [inline graphdatascience/tests/unit/test_gds_sessions.py:455]: missing an assertion

File: graphdatascience/session/gds_sessions.py [hunk 5]
```diff
@@ -82,6 +84,23 @@ def __init__(self, api_credentials: AuraAPICredentials) -> None:
...
+
+        estimation = self._aura_api.estimate_size(node_count, relationship_count, algorithm_categories)
+
+        if estimation.did_exceed_maximum:
+            warnings.warn(
+                f"The estimated memory `{estimation.min_required_memory}` exceeds the maximum size"
+                f" supported of your Aura tenant (`{estimation.recommended_size}`).",
```
Comment (Mats-SX) [inline graphdatascience/session/gds_sessions.py:98]: there's a string interpolation bug in here

Comment (Mats-SX) [reply graphdatascience/session/gds_sessions.py:98]: oh it was fixed in another commit

**Review by Mats-SX — APPROVED** [review]
The SessionMemory allows too small sizes for AuraDS-based sessions. I will reduce its value space in another PR.

File: graphdatascience/session/gds_sessions.py (hunk 3 above)
Comment (Mats-SX) [reply graphdatascience/session/gds_sessions.py:92]: no the 400 was coming from trying to create an AuraDS in Aura API with a size of 4 or 2 or 1

Comment (Mats-SX) [reply graphdatascience/session/gds_sessions.py:92]: yes the Aura API does return only an available size, I think. Although that requires tenant info, which is not there

**Pull request merged**